src/pages/trade/ui/order-form/order-form-store.ts

```diff
@@ -0,0 +1,344 @@
+    );
+    if (baseAssetBalance?.balanceView) {
+      this.baseAsset.setBalanceView(baseAssetBalance.balanceView);
+    }
+    if (baseAssetBalance?.accountAddress) {
+      this.baseAsset.setAccountAddress(baseAssetBalance.accountAddress);
+    }
+
+    const quoteAssetBalance = this.balances?.find(resp =>
+      getAssetIdFromValueView(resp.balanceView).equals(getAssetId(this.quoteAsset.metadata)),
+    );
+    if (quoteAssetBalance?.balanceView) {
+      this.quoteAsset.setBalanceView(quoteAssetBalance.balanceView);
+    }
+    if (quoteAssetBalance?.accountAddress) {
+      this.quoteAsset.setAccountAddress(quoteAssetBalance.accountAddress);
+    }
+  };
+
+  private simulateSwapTx = async (
+    assetIn: OrderFormAsset,
+    assetOut: OrderFormAsset,
+  ): Promise<ValueView> => {
+    const req = new SimulateTradeRequest({
+      input: assetIn.toValue(),
+      output: assetOut.assetId,
+    });
+
+    const res = await penumbra.service(SimulationService).simulateTrade(req);
+
+    const output = new ValueView({
+      valueView: {
+        case: 'knownAssetId',
+        value: {
+          amount: res.output?.output?.amount,
+          metadata: assetOut.metadata,
+        },
+      },
+    });
+
+    return output;
+  };
+
+  setAssets = (baseAsset: Metadata, quoteAsset: Metadata): void => {
+    this.baseAsset = new OrderFormAsset(baseAsset);
+    this.quoteAsset = new OrderFormAsset(quoteAsset);
+
+    this.baseAsset.onAmountChange(this.handleAmountChange);
+    this.quoteAsset.onAmountChange(this.handleAmountChange);
+
+    this.setBalancesOfAssets();
+    void this.calculateGasFee();
+    void this.calculateExchangeRate();
+  };
+
+  setBalances = (balances: BalancesResponse[]): void => {
+    this.balances = balances;
+    this.setBalancesOfAssets();
+  };
+
+  handleAmountChange = async (asset: OrderFormAsset): Promise<void> => {
+    const assetIsBaseAsset = asset.assetId === this.baseAsset.assetId;
+    const assetIn = assetIsBaseAsset ? this.baseAsset : this.quoteAsset;
+    const assetOut = assetIsBaseAsset ? this.quoteAsset : this.baseAsset;
+
+    try {
+      void this.calculateGasFee();
+
+      assetOut.setIsEstimating(true);
+
+      // reset potentially previously set flag
+      assetIn.setIsApproximately(false);
+
+      const output = await this.simulateSwapTx(assetIn, assetOut);
+      const outputAmount = getFormattedAmtFromValueView(output, true);
+
+      assetOut.setAmount(Number(outputAmount), false);
+      assetOut.setIsApproximately(true);
+    } catch (e) {
+      if (e instanceof Error && e.name !== 'PenumbraProviderNotAvailableError') {
+        openToast({
+          type: 'error',
+          message: 'Error estimating amount',
+          description: JSON.stringify(e),
+        });
+      }
+    } finally {
+      assetOut.setIsEstimating(false);
+    }
+  };
+
+  calculateExchangeRate = async (): Promise<void> => {
```

Comment:
question: for now, it always returns 0 for me in UM/USDC. Should it be like this or is it different in the environment where you tested?

Reply:
I think it's a problem for this pair on testnet